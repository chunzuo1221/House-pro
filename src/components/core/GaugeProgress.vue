<template>
<div class="gauge-progress" percentage="60">
  <svg :id="id" class="stat-circle" viewBox="0 0 36 45">
    <path class="progress-bg"
      d="M18 33.9155
      a 15.9155 15.9155 0 0 1 0 -31.831
      a 15.9155 15.9155 0 0 1 0 31.831"
    />
    <path v-bind:class="{ progress: isActive }"
      d="M18 33.9155
      a 15.9155 15.9155 0 0 1 0 -31.831
      a 15.9155 15.9155 0 0 1 0 31.831"
    />
    <image :href="image" x="12" y="12" height="12px" width="12px"/>
    <text :x="xOffSet1" y="95%" v-show="maxValue !== 0" fill="#528EC1" font-size="12" font-weight="bold">{{ value }}</text>
    <text v-show="hideMax !== true && maxValue !== 0" :x="xOffSet2" y="95%" fill="#AAA" font-size="4">/{{ maxValue }}</text>
  </svg>
</div>
</template>
<script>
export default {
  name: 'GaugeProgress',
  props: {
    id: {
      type: String,
      required: true,
      default: null
    },
    value: {
      type: Number,
      required: true,
      default: 0.0
    },
    maxValue: {
      type: Number,
      required: true,
      default: 0.0
    },
    image: {
      type: String,
      required: false,
      default: '/static/images/house_icon.png'
    },
    hideMax: {
      type: Boolean,
      required: false,
      default: false
    },
    height: {
      type: Number,
      required: false,
      default: 170
    }
  },
  computed: {
    xOffSet1 () {
      var value = '42%'
      if (this.value.toString().length > 1) {
        value = '30%'
      }
      if (this.value.toString().length > 3) {
        value = '14%'
      }
      return value
    },
    xOffSet2 () {
      var value = '65%'
      if (this.value.toString().length > 1) {
        value = '78%'
      }
      if (this.value.toString().length > 3) {
        value = '82%'
      }
      return value
    }
  },
  data () {
    return {
      isActive: false
    }
  },
  mounted () {
    this.updateUI()
  },
  methods: {
    updateUI () {
      if (this.maxValue > 0) {
        this.isActive = false
        let calc = (this.value * (0.8) / (this.maxValue / 100)).toFixed(0)
        calc = calc > 80 ? 80 : calc
        this.$el.style.setProperty('--progress', calc)
        setTimeout(() => {
          this.isActive = true
          this.$forceUpdate()
        }, 100)
      }
      if (document.getElementById(this.id)) {
        document.getElementById(this.id).style.height = this.height + 'px'
      }
    }
  },
  watch: {
    value: function (val) {
      this.updateUI()
    },
    maxValue: function (val) {
      this.updateUI()
    },
    image: function (val) {
      this.updateUI()
    }
  }
}
</script>

<style scoped>
:root {
  --progress: 0;
}
body {
  padding: 50px;
}
.stat-circle path {
  fill: none;
}
.stat-circle path.progress-bg {
  stroke: #e0dfdf;
  stroke-width: 3.5;
  stroke-dasharray: 80 100;
  stroke-dashoffset: -10;
}
.stat-circle path.progress {
  stroke: #528EC1;
  stroke-width: 3.5;
  stroke-dashoffset: -10;
  transition-timing-function: cubic-bezier(.29, 1.01, 1, -0.68);
  animation: dash 2s linear forwards;
}
@keyframes dash {
  from {
    stroke-dasharray: 0 100;
  }
  to {
    stroke-dasharray: var(--progress) 100;
  }
}
</style>
