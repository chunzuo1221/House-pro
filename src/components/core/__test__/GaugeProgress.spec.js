import { shallowMount } from '@vue/test-utils'
import GaugeProgress from '@/components/core/GaugeProgress'

describe('GaugeProgress', () => {
  it('renders correctly', () => {
    const wrapper = shallowMount(GaugeProgress, {
      propsData: {
        id: 'gauge_progress',
        value: -1,
        maxValue: 10,
        image: null
      }
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
