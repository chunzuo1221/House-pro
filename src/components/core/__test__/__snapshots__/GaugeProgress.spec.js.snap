// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GaugeProgress renders correctly 1`] = `
<div
  class="gauge-progress"
  percentage="60"
>
  <svg
    class="stat-circle"
    id="gauge_progress"
    viewBox="0 0 36 45"
  >
    <path
      class="progress-bg"
      d="M18 33.9155
      a 15.9155 15.9155 0 0 1 0 -31.831
      a 15.9155 15.9155 0 0 1 0 31.831"
    />
     
    <path
      class=""
      d="M18 33.9155
      a 15.9155 15.9155 0 0 1 0 -31.831
      a 15.9155 15.9155 0 0 1 0 31.831"
    />
     
    <image
      height="12px"
      width="12px"
      x="12"
      y="12"
    />
     
    <text
      fill="#528EC1"
      font-size="12"
      font-weight="bold"
      x="30%"
      y="95%"
    >
      -1
    </text>
     
    <text
      fill="#AAA"
      font-size="4"
      x="78%"
      y="95%"
    >
      /10
    </text>
  </svg>
</div>
`;
